### ISSUE TITLE
Question about decoupling

### ISSUE THREAD

**Issue by robinlehrmann** [patchlevel/event-sourcing#298]
Hi :)

I want to ask you @DavidBadura and @DanielBadura if you have an idea how I can decouple your library from my Domain code.
In a private project I use a hexagonal architecture approche and my model extends from `Patchlevel\EventSourcing\Aggregate\AggregateRoot` in the Domain namespace.

Do you have an idea how I can leave the events and the aggregate in the Domain code and everything which depends on your lib in Infrastructure ?

I'm currently thinking about the proxy classes of doctrine/orm. Imagine we configure a model as an aggregateRoot and the configuration is just in yaml. To have the inheritance of `Patchlevel\EventSourcing\Aggregate\AggregateRoot` with the model you could create proxy classes which extends then from that base class and the domain could does not know anything about the patchlevel/eventsourcing library, what do you think? :)

Feel free to contact me if you want to know more about what I'm thinking about or just write here if you have enough information :)

**Comment by DanielBadura** [comment]
For sure we want to enable the user to decouple as much as possible from the library. But here we are talking about the logic that needs to be done in the aggregate which we can't push to the user. We need some logic in the aggregate to properly handle the events in the right manner. You linked the other ticket where we thought about providing an interface and a trait instead of abstract class. But with that the same problem is there - you are still depending on the lib the difference is the symbol. IMHO it is not really possible for us to provide you no direct dependency in your domain folder.

**Comment by robinlehrmann** [comment]
Hi @DanielBadura
but could it not be possible to generate a proxy class like doctrine/orm does with it's `Proxies\__CG__\` namespaced classes?

I thought about something like this:
```
class SomeAggregate
{
    private array $uncommitedEvents = [];

    public static function create(): self
    {
        // do something and create event
        $this->uncommitedEvents[] = $event;
    }
}
```

class in cache:
```
class ProxyedSomeAggregate extends AggregateRoot
{
    // everything here is related to your code of the lib
}
```

Not sure if this is understandable 😅

**Comment by DavidBadura** [comment]
The AggregateRoot has the following logic and needs it to work:

1) record new events (with the next playhead)
2) return new recorded events
3) increment the version of the aggregate (playhead)
4) find an apply method that must be executed immediately after the record
5) execute the apply method
6) when rebuilding, repeat step 3-5

We can provide an interface that defines the API, how to read out the events for the aggregate and how to rebuild the aggregate. But then it still remains that the user has to build the entire aggregate logic himself. We can only improve that with a trait at most. This is the only "separation" we can offer.

Doctrine proxies have a completely different purpose and have nothing to do with being separated from the library. The proxy classes are only used for lazy loading. If you have a "user" that has a reference to a "group", and you make it available through a getter. Then a proxy class is returned when called, so that the group is not loaded directly. Only when you execute a method on the group (except for an id), the data will be loaded so that the real value can be returned.

Btw. doctrine also forces you to use some things that you can't get from your entities. For example: ArrayCollection. This is necessary so that the references can be loaded lazy. You can't get rid of it.

In summary: Sometimes you can't avoid a coupling, but of course the aim is to keep it as small as possible, see ticket #245. Proxy classes cannot help here and mostly only have use for lazy loading.

**Comment by robinlehrmann** [comment]
Thank you for the good explanation @DavidBadura and @DanielBadura also for the answer :)
I guess then the trait is the better approche 👍

**Comment by DanielBadura** [comment]
Right now we see no benefit in switching to an interface and trait solution. So we are sticking to the abstract class as long as it stays like this.

**Comment by PonendoPonens** [comment]
Hello,

There is argument to switch to traits. I found EventSauce to be very flexible  when using traits and interfaces together. First - you can remove inheritance which has no deeper sense here - you have 0 benefits from keeping inheritance which is only ball and chain for many developers. Second - you can add more optional functionalities to aggregates which i found usefull in the past. Like nesting aggregates to keep aggregateRoot more lean. Some commands are simply passed to sub-aggregates which can produce their own events. They are still part of one aggregateRoot, but from logical point of view, they care their internal logic (like managing collections of VO. CartItems, Photos etc.). Without flexibility, AR's became big and messy. 

https://eventsauce.io/docs/advanced/aggregate-root-with-aggregates/

I like your approach a little more, because of tools you provide, but inheritance is definetly limiting in this particular case. In the fact - i can't move my domain models without additional boilerplate code around.

**Comment by DavidBadura** [comment]
Hello @PonendoPonens, thanks for sharing your opinion! We have a ticket open for this and will look at it again: https://github.com/patchlevel/event-sourcing/issues/245

**Comment by DanielBadura** [comment]
@PonendoPonens can you have a look here at this PoC https://github.com/patchlevel/event-sourcing/pull/351 if the interface is sufficient for your case? Also, could please go a little more into detail with nesting the aggregate roots?
